File: backend/app/schemas/investigation.py
"""
Schemas for Graph-RAG fraud investigations.
"""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, Field
from app.graph.query_models import (
    EntityRiskResponse,
    FraudRingResponse,
    GraphNeighborsResponse,
    RelatedIncidentsResponse,
    SharedEntityResponse,
)


class InvestigationTargetType(str, Enum):
    """
    Supported investigation target types.
    """

    COMPLAINT = "complaint"
    PHONE = "phone"
    EMAIL = "email"
    UPI = "upi"
    BANK_ACCOUNT = "bank_account"
    DEVICE = "device"
    IP = "ip"
    PERSON = "person"


class InvestigationRequest(BaseModel):
    """
    Request for generating an investigation report.
    """

    target_type: InvestigationTargetType = Field(
        ...,
        description="Type of entity being investigated.",
    )

    target_value: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Identifier or value of the investigation target.",
    )


class ConnectedEntity(BaseModel):
    """
    Connected entity discovered during graph retrieval.
    """

    entity_type: str
    entity_value: str
    relationship: str


class InvestigationEvidence(BaseModel):
    """
    Complete graph evidence collected for an investigation.
    This object acts as the Graph-RAG retrieval context.
    """

    neighbors: GraphNeighborsResponse

    related_incidents: RelatedIncidentsResponse

    risk: EntityRiskResponse

    fraud_ring: FraudRingResponse

    shared_entities: SharedEntityResponse


class InvestigationReport(BaseModel):
    """
    AI-generated investigation report.
    """

    summary: str

    risk_level: str

    confidence: float

    findings: list[str]

    key_entities: list[str]

    recommended_actions: list[str]


class InvestigationResponse(BaseModel):
    """
    Complete investigation response.
    """

    target_type: InvestigationTargetType

    target_value: str

    evidence: InvestigationEvidence

    report: InvestigationReport

File: frontend/src/types/investigation.ts
import {
  EntityRiskResponse,
  FraudRingResponse,
  GraphNeighborsResponse,
  RelatedIncidentsResponse,
  SharedEntityResponse,
} from './graph';

export type InvestigationTargetType =
  | 'complaint'
  | 'phone'
  | 'email'
  | 'upi'
  | 'bank_account'
  | 'device'
  | 'ip'
  | 'person';

export interface InvestigationRequest {
  target_type: InvestigationTargetType;
  target_value: string;
}

export interface InvestigationEvidence {
  neighbors: GraphNeighborsResponse;
  related_incidents: RelatedIncidentsResponse;
  risk: EntityRiskResponse;
  fraud_ring: FraudRingResponse;
  shared_entities: SharedEntityResponse;
}

export interface InvestigationReport {
  summary: string;
  risk_level: string;
  confidence: number;
  findings: string[];
  key_entities: string[];
  recommended_actions: string[];
}

export interface InvestigationResponse {
  target_type: InvestigationTargetType;
  target_value: string;
  evidence: InvestigationEvidence;
  report: InvestigationReport;
}

export interface ReportTelemetry {
  correlation_id: string;
  provider: string;
  model: string;
  latency_ms: number;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  prompt_hash: string;
  template_version: string;
  summary_version: string;
  report_context_version: string;
  generated_at: string;
}

export interface ProfessionalInvestigationReport {
  report_id: string;
  target_value: string;
  generated_at: string;
  executive_summary: {
    summary_text: string;
    overall_risk_level: string;
    key_takeaways: string[];
  };
  investigation_scope: {
    target_value: string;
    target_type?: string | null;
    total_complaints: number;
    total_entities: number;
    duration_days: number;
  };
  timeline_summary: {
    timeline_narrative: string;
    milestones: Array<{
      event_type: string;
      timestamp: string;
      title: string;
      description: string;
    }>;
  };
  key_findings: Array<{
    finding_id: string;
    title: string;
    description: string;
    severity: string;
    confidence: number;
    citations: string[];
  }>;
  fraud_network_evolution: {
    evolution_narrative: string;
    network_stage: string;
  };
  evidence_assessment: {
    evidence_summary: string;
    supporting_evidence_count: number;
  };
  recommendations: Array<{
    recommendation_id: string;
    action: string;
    priority: string;
    rationale: string;
    trigger: string;
    target_entities: string[];
  }>;
  limitations: {
    data_quality_assessment: string;
    limitations: string[];
  };
  conclusion: {
    summary_conclusion: string;
  };
  telemetry: ReportTelemetry;
}
